### PR TITLE
Fix where Datanode crash due to oom when delta buff accumulating constantly because of non auto flush

### PR DESCRIPTION
**Pull request by MrPresent-Han**
enable delbuf auto flush function to avoid OOM when del msg accumulating

Signed-off-by: MrPresent-Han <jamesharden11122@gmail.com>

#19713

### CODE REVIEW

File: internal/datanode/buffer.go [hunk 1]
```diff
@@ -17,13 +17,220 @@
...
 	"errors"
+	"fmt"
+	"github.com/milvus-io/milvus-proto/go-api/schemapb"
+	"github.com/milvus-io/milvus/internal/metrics"
 	"math"
+	"sync"
 
 	"github.com/milvus-io/milvus/internal/proto/datapb"
 	"github.com/milvus-io/milvus/internal/storage"
 )
```
Comment (MrPresent-Han) [inline internal/datanode/buffer.go:31]: using buf manager to replace sync.map, which is too weak to maintain related data in the memory

File: internal/datanode/buffer.go [hunk 2]
```diff
@@ -17,13 +17,220 @@
...
+// DelBufferManager is in charge of managing insertBuf and delBuf from an overall prospect
+// not only controlling buffered data size based on every segment size, but also triggering
+// insert/delete flush when the memory usage of the whole manager reach a certain level.
+// but at the first stage, this struct is only used for delete buff
+type DelBufferManager struct {
+	delBufMap     sync.Map // map[segmentID]*DelDataBuf
+	delMemorySize int64
+	delBufHeap    *PriorityQueue
+}
+
```
Comment (MrPresent-Han) [inline internal/datanode/buffer.go:41]: use priorityQueue to mange delDataBuf and the amount of memory used by every delDataBuf, in order to identify
correct segments set to flush for auto flushing

File: internal/datanode/buffer.go [hunk 3]
```diff
@@ -57,10 +264,11 @@ func (bd *BufferData) updateTimeRange(tr TimeRange) {
 type DelDataBuf struct {
 	datapb.Binlog
 	delData *DeleteData
+	item    *Item
 }
```
Comment (MrPresent-Han) [inline internal/datanode/buffer.go:268]: as entriesNum inside DelDataBuf will only increase and never decrease, I modify the function name here to 'accumulate'

File: internal/datanode/compactor_test.go [hunk 4]
```diff
@@ -853,6 +855,9 @@ func (mfm *mockFlushManager) flushDelData(data *DelDataBuf, segmentID UniqueID,
 	if mfm.returnError {
 		return fmt.Errorf("mock error")
 	}
```
Comment (MrPresent-Han) [inline internal/datanode/compactor_test.go:857]: add recordFlushedSeg to record all flushed segments in the process of running ut, aiming at verifying the related code is running correctly

File: internal/util/paramtable/component_param.go [hunk 5]
```diff
@@ -1396,6 +1398,16 @@ func (p *dataNodeConfig) initFlushInsertBufferSize() {
 	p.FlushInsertBufferSize = bs
 }
```
Comment (MrPresent-Han) [inline internal/util/paramtable/component_param.go:1400]: the config point of delete buffer size, the default confined value here is 64MB per channel
The value may be too big for production environment

File: internal/datanode/buffer.go [hunk 6]
```diff
@@ -17,13 +17,220 @@
...
+type Item struct {
+	segmentID  UniqueID // The segmentID
+	memorySize int64    // The size of memory consumed by del buf
+	index      int      // The index of the item in the heap.
+	// The index is needed by update and is maintained by the heap.Interface methods.
+}
+
+// A PriorityQueue implements heap.Interface and holds Items.
+// We use PriorityQueue to manage memory consumed by del buf
+type PriorityQueue []*Item
```
Comment (jiaoew1991) [inline internal/datanode/buffer.go:196]: there is already a `PriorityQueue` in `indexcoord`, should move these 2 types in util and reuse

File: internal/datanode/buffer.go [hunk 7]
```diff
@@ -17,13 +17,220 @@
...
+// not only controlling buffered data size based on every segment size, but also triggering
+// insert/delete flush when the memory usage of the whole manager reach a certain level.
+// but at the first stage, this struct is only used for delete buff
+type DelBufferManager struct {
+	delBufMap     sync.Map // map[segmentID]*DelDataBuf
+	delMemorySize int64
+	delBufHeap    *PriorityQueue
+}
+
+func (bm *DelBufferManager) SegDelBufMemSize(segID UniqueID) int64 {
```
Comment (jiaoew1991) [inline internal/datanode/buffer.go:42]: GetDelBufMemSize?

File: internal/datanode/buffer.go [hunk 8]
```diff
@@ -17,13 +17,220 @@
...
+
+// Store :the method only for unit test
+func (bm *DelBufferManager) Store(segID UniqueID, delDataBuf *DelDataBuf) {
+	bm.delBufMap.Store(segID, delDataBuf)
+}
+
+func (bm *DelBufferManager) StoreNewDeletes(segID UniqueID, pks []primaryKey,
+	tss []Timestamp, tr TimeRange) {
+	//1. load or create delDataBuf
+	var delDataBuf *DelDataBuf
```
Comment (jiaoew1991) [inline internal/datanode/buffer.go:64]: all of the functions are not protected by a lock, is that right?

Comment (MrPresent-Han) [reply internal/datanode/buffer.go:64]: According to Congqi, all operations related to del/insert buf will be executed inside one channel and one groutine, so they will be run in serial sequence

File: internal/datanode/buffer.go (hunk 7 above)
Comment (MrPresent-Han) [reply internal/datanode/buffer.go:42]: changed

**Comment by MrPresent-Han** [comment]
rerun ut

File: internal/datanode/buffer.go [hunk 9]
```diff
@@ -17,13 +17,220 @@
...
+	memorySize int64    // The size of memory consumed by del buf
+	index      int      // The index of the item in the heap.
+	// The index is needed by update and is maintained by the heap.Interface methods.
+}
+
+// A PriorityQueue implements heap.Interface and holds Items.
+// We use PriorityQueue to manage memory consumed by del buf
+type PriorityQueue []*Item
+
+func (pq PriorityQueue) Len() int { return len(pq) }
```
Comment (soothing-rain) [inline internal/datanode/buffer.go:198]: here and below:

func (pq *PriorityQueue)

File: internal/datanode/buffer.go [hunk 10]
```diff
@@ -17,13 +17,220 @@
...
+type Item struct {
+	segmentID  UniqueID // The segmentID
+	memorySize int64    // The size of memory consumed by del buf
+	index      int      // The index of the item in the heap.
+	// The index is needed by update and is maintained by the heap.Interface methods.
+}
+
+// A PriorityQueue implements heap.Interface and holds Items.
+// We use PriorityQueue to manage memory consumed by del buf
+type PriorityQueue []*Item
```
Comment (soothing-rain) [inline internal/datanode/buffer.go:196]: Maybe

type PriorityQueue struct {
	items []*Item
}

File: internal/datanode/buffer.go [hunk 11]
```diff
@@ -17,13 +17,220 @@
...
+// Store :the method only for unit test
+func (bm *DelBufferManager) Store(segID UniqueID, delDataBuf *DelDataBuf) {
+	bm.delBufMap.Store(segID, delDataBuf)
+}
+
+func (bm *DelBufferManager) StoreNewDeletes(segID UniqueID, pks []primaryKey,
+	tss []Timestamp, tr TimeRange) {
+	//1. load or create delDataBuf
+	var delDataBuf *DelDataBuf
+	value, loaded := bm.delBufMap.Load(segID)
```
Comment (soothing-rain) [inline internal/datanode/buffer.go:65]: if value, loaded := bm.delBufMap.Load(segID) ; loaded {
		delDataBuf = value.(*DelDataBuf)
	} else {
		delDataBuf = newDelDataBuf()
	}

File: internal/datanode/buffer.go [hunk 12]
```diff
@@ -17,13 +17,220 @@
...
+		delData.Pks = append(delData.Pks, pks[i])
+		delData.Tss = append(delData.Tss, tss[i])
+		switch pks[i].Type() {
+		case schemapb.DataType_Int64:
+			bufSize += 8
+		case schemapb.DataType_VarChar:
+			varCharPk := pks[i].(*varCharPrimaryKey)
+			bufSize += int64(len(varCharPk.Value))
+		}
+		bufSize += 8
```
Comment (soothing-rain) [inline internal/datanode/buffer.go:91]: Is '+= 8' this for timestamp? Better add comment for better readability.

File: internal/datanode/buffer.go [hunk 13]
```diff
@@ -17,13 +17,224 @@
...
+func (bm *DelBufferManager) Load(segID UniqueID) (delDataBuf *DelDataBuf, ok bool) {
+	value, ok := bm.delBufMap.Load(segID)
+	if ok {
+		return value.(*DelDataBuf), ok
+	}
+	return nil, ok
+}
+
+func (bm *DelBufferManager) Delete(segID UniqueID) {
+	buf, ok := bm.delBufMap.Load(segID)
```
Comment (soothing-rain) [inline internal/datanode/buffer.go:121]: -->
if buf, ok := bm.delBufMap.Load(segID); ok {
...
}

File: internal/datanode/buffer.go [hunk 14]
```diff
@@ -17,13 +17,224 @@
...
+	if ok {
+		delDataBuf := buf.(*DelDataBuf)
```
Comment (soothing-rain) [inline internal/datanode/buffer.go:123]: -->

item := buf.(*DelDataBuf).item

File: internal/datanode/buffer.go [hunk 15]
```diff
@@ -17,13 +17,224 @@
...
+		item := delDataBuf.item
+		bm.delMemorySize -= item.GetMemorySize()
+		heap.Remove(bm.delBufHeap, item.GetIndex())
+		bm.delBufMap.Delete(segID)
+	}
+}
+
+func (bm *DelBufferManager) LoadAndDelete(segID UniqueID) (delDataBuf *DelDataBuf, ok bool) {
+	buf, ok := bm.Load(segID)
```
Comment (soothing-rain) [inline internal/datanode/buffer.go:132]: ditto

File: internal/datanode/buffer.go [hunk 16]
```diff
@@ -17,13 +17,224 @@
...
+// not only controlling buffered data size based on every segment size, but also triggering
+// insert/delete flush when the memory usage of the whole manager reach a certain level.
+// but at the first stage, this struct is only used for delete buff
+type DelBufferManager struct {
+	delBufMap     sync.Map // map[segmentID]*DelDataBuf
+	delMemorySize uint64
+	delBufHeap    *PriorityQueue
+}
+
+func (bm *DelBufferManager) GetSegDelBufMemSize(segID UniqueID) uint64 {
```
Comment (soothing-rain) [inline internal/datanode/buffer.go:47]: here and below: There's not need to capitalize if the method is exported.

File: internal/datanode/buffer.go [hunk 17]
```diff
@@ -17,13 +17,220 @@
...
+	}
+}
+
+func (bm *DelBufferManager) ShouldFlushSegments() []UniqueID {
+	var shouldFlushSegments []UniqueID
+	if bm.delMemorySize < Params.DataNodeCfg.FlushDeleteBufferBytes {
+		return shouldFlushSegments
+	} else {
+		mmUsage := bm.delMemorySize
+		var popedSegMem []*Item
```
Comment (soothing-rain) [inline internal/datanode/buffer.go:167]: nit: popped

File: internal/datanode/buffer.go [hunk 18]
```diff
@@ -17,13 +17,220 @@
...
+		for {
+			segMem := heap.Pop(bm.delBufHeap).(*Item)
+			popedSegMem = append(popedSegMem, segMem)
+			shouldFlushSegments = append(shouldFlushSegments, segMem.segmentID)
```
Comment (soothing-rain) [inline internal/datanode/buffer.go:171]: add some logs around here.

File: internal/datanode/flow_graph_delete_node.go [hunk 19]
```diff
@@ -119,27 +115,46 @@ func (dn *deleteNode) Operate(in []Msg) []Msg {
 		dn.showDelBuf(segIDs, fgMsg.timeRange.timestampMax)
 	}
 
+	//here we adopt a quite radical strategy:
+	//every time we make sure that the N biggest delDataBuf can be flushed
+	//when memsize usage reaches a certain level
+	//then we will add all segments in the fgMsg.segmentsToFlush into the toFlushSeg and remove duplicate segments
+	//the aim for taking all these actions is to guarantee that the memory consumed by delBuf will not exceed a limit
+	segmentsToFlush := dn.delBufferManager.ShouldFlushSegments()
```
Comment (soothing-rain) [inline internal/datanode/flow_graph_delete_node.go:123]: add some logs to catch the `ShouldFlushSegments` list

File: internal/util/paramtable/component_param.go [hunk 20]
```diff
@@ -1396,6 +1398,16 @@ func (p *dataNodeConfig) initFlushInsertBufferSize() {
...
 
+func (p *dataNodeConfig) initFlushDeleteBufferSize() {
+	delBufferBytesStr := p.Base.LoadWithDefault2([]string{"DATA_NODE_DELETE_BUFFER_BYTES",
+		"datanode.flush.deleteBufBytes"}, "67108864")
```
Comment (soothing-rain) [inline internal/util/paramtable/component_param.go:1403]: write it as 64 * 1024 * 1024, also please add datanode.flush.deleteBufBytes config to milvus.yaml

File: internal/datanode/buffer.go (hunk 13 above)
Comment (MrPresent-Han) [reply internal/datanode/buffer.go:116]: changed

File: internal/datanode/buffer.go (hunk 17 above)
Comment (MrPresent-Han) [reply internal/datanode/buffer.go:167]: changed

File: internal/datanode/buffer.go (hunk 18 above)
Comment (MrPresent-Han) [reply internal/datanode/buffer.go:171]: added

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by soothing-rain** [comment]
Please update the title of your PR to explain what you have fixed.

/lgtm

**Pull request merged**